File: pitchcraft-ai/services/firecrawlService.js
const fetch = require('node-fetch');

const FIRECRAWL_URL = process.env.FIRECRAWL_URL || 'https://api.firecrawl.dev/v0';
const API_KEY = process.env.FIRECRAWL_API_KEY;

const researchIdea = async (idea) => {
  try {
    if (!API_KEY) {
      console.warn('Firecrawl API key not found, using fallback research');
      return generateFallbackResearch(idea);
    }

    // Search for market information
    const searchQueries = [
      `${idea} market size trends 2024`,
      `${idea} competitors analysis`,
      `${idea} industry report statistics`,
      `${idea} startup funding investors`
    ];

    const researchPromises = searchQueries.map(query => searchMarketData(query));
    const results = await Promise.allSettled(researchPromises);
    
    const successfulResults = results
      .filter(result => result.status === 'fulfilled' && result.value)
      .map(result => result.value);

    if (successfulResults.length === 0) {
      console.warn('No research data found, using fallback');
      return generateFallbackResearch(idea);
    }

    return {
      marketSize: extractMarketSize(successfulResults),
      competitors: extractCompetitors(successfulResults),
      trends: extractTrends(successfulResults),
      funding: extractFundingInfo(successfulResults),
      sources: successfulResults.map(r => r.url).filter(Boolean)
    };

  } catch (error) {
    console.error('Firecrawl research error:', error.message);
    return generateFallbackResearch(idea);
  }
};

const searchMarketData = async (query) => {
  try {
    const response = await fetch(`${FIRECRAWL_URL}/search`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${API_KEY}`
      },
      body: JSON.stringify({
        query: query,
        pageOptions: {
          onlyMainContent: true
        },
        limit: 5
      })
    });

    if (!response.ok) {
      console.error(`Firecrawl search error for "${query}":`, response.status);
      return null;
    }

    const data = await response.json();
    return data.data?.[0]; // Return first result

  } catch (error) {
    console.error(`Search error for "${query}":`, error.message);
    return null;
  }
};

const extractMarketSize = (results) => {
  // Look for market size indicators in the research data
  const marketKeywords = ['billion', 'million', 'market size', 'valuation', 'revenue'];
  
  for (const result of results) {
    const content = result.content || result.text || '';
    const lowerContent = content.toLowerCase();
    
    if (marketKeywords.some(keyword => lowerContent.includes(keyword))) {
      // Extract relevant market size information
      const sentences = content.split('.').filter(sentence => 
        marketKeywords.some(keyword => sentence.toLowerCase().includes(keyword))
      );
      
      if (sentences.length > 0) {
        return sentences[0].trim() + '.';
      }
    }
  }
  
  return 'Large and growing market with significant opportunity for expansion.';
};

const extractCompetitors = (results) => {
  // Look for competitor mentions
  const competitors = [];
  const competitorKeywords = ['competitor', 'rival', 'alternative', 'similar', 'leading company'];
  
  for (const result of results) {
    const content = result.content || result.text || '';
    const lowerContent = content.toLowerCase();
    
    if (competitorKeywords.some(keyword => lowerContent.includes(keyword))) {
      // Extract company names (simplified approach)
      const sentences = content.split('.').filter(sentence => 
        competitorKeywords.some(keyword => sentence.toLowerCase().includes(keyword))
      );
      
      competitors.push(...sentences.slice(0, 2));
    }
  }
  
  return competitors.length > 0 ? competitors : [
    'Established players with traditional approaches',
    'Emerging startups with innovative solutions',
    'Large enterprises with comprehensive platforms'
  ];
};

const extractTrends = (results) => {
  const trendKeywords = ['trend', 'growth', 'increasing', 'rising', 'adoption', 'emerging'];
  const trends = [];
  
  for (const result of results) {
    const content = result.content || result.text || '';
    const lowerContent = content.toLowerCase();
    
    if (trendKeywords.some(keyword => lowerContent.includes(keyword))) {
      const sentences = content.split('.').filter(sentence => 
        trendKeywords.some(keyword => sentence.toLowerCase().includes(keyword))
      );
      
      trends.push(...sentences.slice(0, 2));
    }
  }
  
  return trends.length > 0 ? trends : [
    'Digital transformation driving market growth',
    'Increasing demand for innovative solutions',
    'Growing adoption of technology-based approaches'
  ];
};

const extractFundingInfo = (results) => {
  const fundingKeywords = ['funding', 'investment', 'raised', 'venture', 'series', 'capital'];
  const fundingInfo = [];
  
  for (const result of results) {
    const content = result.content || result.text || '';
    const lowerContent = content.toLowerCase();
    
    if (fundingKeywords.some(keyword => lowerContent.includes(keyword))) {
      const sentences = content.split('.').filter(sentence => 
        fundingKeywords.some(keyword => sentence.toLowerCase().includes(keyword))
      );
      
      fundingInfo.push(...sentences.slice(0, 2));
    }
  }
  
  return fundingInfo.length > 0 ? fundingInfo : [
    'Active investor interest in the sector',
    'Multiple funding rounds completed by similar companies',
    'Growing venture capital investment in the space'
  ];
};

const generateFallbackResearch = (idea) => {
  return {
    marketSize: `The ${idea} market represents a significant opportunity with strong growth potential and increasing demand.`,
    competitors: [
      'Established market leaders with traditional approaches',
      'Innovative startups with technology-focused solutions',
      'Large enterprises with comprehensive service offerings'
    ],
    trends: [
      'Growing adoption of digital solutions',
      'Increasing market demand and customer awareness',
      'Technology-driven transformation in the industry'
    ],
    funding: [
      'Active venture capital interest in the sector',
      'Multiple successful funding rounds by competitors',
      'Growing investor confidence in market potential'
    ],
    sources: []
  };
};

module.exports = {
  researchIdea
};

File: pitchcraft-new/services/firecrawl_service.py
import httpx
import os
import json
from typing import Dict, List, Optional
from datetime import datetime

class FirecrawlService:
    def __init__(self):
        self.api_key = os.getenv("FIRECRAWL_API_KEY")
        self.base_url = os.getenv("FIRECRAWL_URL", "https://api.firecrawl.dev")
        
    async def check_health(self) -> bool:
        """Check if Firecrawl service is available"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/v0/status",
                    headers={"Authorization": f"Bearer {self.api_key}"},
                    timeout=10.0
                )
                return response.status_code == 200
        except:
            return False
    
    async def test_connection(self) -> Dict:
        """Test Firecrawl API connection"""
        try:
            async with httpx.AsyncClient() as client:
                # Test with a simple scrape
                response = await client.post(
                    f"{self.base_url}/v0/scrape",
                    headers={
                        "Authorization": f"Bearer {self.api_key}",
                        "Content-Type": "application/json"
                    },
                    json={
                        "url": "https://firecrawl.dev/",
                        "pageOptions": {
                            "onlyMainContent": True
                        }
                    },
                    timeout=15.0
                )
                
                if response.status_code == 200:
                    result = response.json()
                    return {
                        "status": "connected",
                        "message": "Firecrawl API is working",
                        "data": result.get("data", {}).get("content", "")[:100] + "..."
                    }
                else:
                    return {
                        "status": "error",
                        "code": response.status_code,
                        "message": response.text
                    }
                    
        except Exception as e:
            return {
                "status": "error",
                "message": f"Connection failed: {str(e)}"
            }
    
    async def research_startup_idea(self, idea: str, industry: Optional[str] = None) -> Dict:
        """Research startup idea using web scraping and search"""
        try:
            # Generate search queries for market research
            search_queries = self._generate_search_queries(idea, industry)
            research_results = {}
            
            # Research each query
            for query_type, query in search_queries.items():
                try:
                    # Use search functionality to find relevant URLs
                    search_result = await self._search_and_scrape(query)
                    research_results[query_type] = search_result
                except Exception as e:
                    print(f"Error researching {query_type}: {str(e)}")
                    research_results[query_type] = self._generate_fallback_research(query_type, idea)
            
            # Compile comprehensive research data
            compiled_research = self._compile_research_data(research_results, idea)
            return compiled_research
            
        except Exception as e:
            print(f"Error in startup research: {str(e)}")
            return self._generate_fallback_comprehensive_research(idea, industry)
    
    async def _search_and_scrape(self, query: str) -> Dict:
        """Search for URLs and scrape relevant content"""
        try:
            async with httpx.AsyncClient() as client:
                # Use Firecrawl's search functionality
                response = await client.post(
                    f"{self.base_url}/v0/search",
                    headers={
                        "Authorization": f"Bearer {self.api_key}",
                        "Content-Type": "application/json"
                    },
                    json={
                        "query": query,
                        "pageOptions": {
                            "onlyMainContent": True
                        },
                        "limit": 3
                    },
                    timeout=20.0
                )
                
                if response.status_code == 200:
                    result = response.json()
                    search_data = result.get("data", [])
                    
                    # Extract key information from search results
                    processed_data = []
                    for item in search_data[:3]:  # Limit to top 3 results
                        processed_data.append({
                            "url": item.get("url", ""),
                            "title": item.get("title", ""),
                            "content": item.get("content", "")[:500] + "...",  # Truncate content
                            "relevance": "high"
                        })
                    
                    return {
                        "query": query,
                        "results": processed_data,
                        "total_found": len(processed_data)
                    }
                else:
                    return self._generate_fallback_search_result(query)
                    
        except Exception as e:
            print(f"Error in search and scrape: {str(e)}")
            return self._generate_fallback_search_result(query)
    
    def _generate_search_queries(self, idea: str, industry: Optional[str] = None) -> Dict[str, str]:
        """Generate relevant search queries for market research"""
        base_industry = industry or "technology"
        
        return {
            "market_size": f"{idea} market size {base_industry} 2024",
            "competitors": f"{idea} competitors startup companies {base_industry}",
            "trends": f"{idea} industry trends {base_industry} innovation",
            "funding": f"{idea} startup funding investment {base_industry}",
            "customers": f"{idea} target customers user personas {base_industry}"
        }
    
    def _compile_research_data(self, research_results: Dict, idea: str) -> Dict:
        """Compile all research results into structured data"""
        compiled = {
            "idea": idea,
            "research_timestamp": datetime.now().isoformat(),
            "market_intelligence": {},
            "competitive_landscape": {},
            "industry_trends": {},
            "funding_landscape": {},
            "customer_insights": {},
            "summary": {}
        }
        
        # Process market size data
        if "market_size" in research_results:
            market_data = research_results["market_size"]
            compiled["market_intelligence"] = {
                "size_indicators": self._extract_market_insights(market_data),
                "growth_potential": "High based on current trends",
                "key_segments": ["B2B", "B2C", "Enterprise"]
            }
        
        # Process competitive data
        if "competitors" in research_results:
            competitor_data = research_results["competitors"]
            compiled["competitive_landscape"] = {
                "direct_competitors": self._extract_competitors(competitor_data),
                "competitive_advantages": ["Innovation", "Technology", "User Experience"],
                "market_gaps": ["Scalability", "Affordability", "Ease of use"]
            }
        
        # Process trends
        if "trends" in research_results:
            trends_data = research_results["trends"]
            compiled["industry_trends"] = {
                "emerging_trends": self._extract_trends(trends_data),
                "technological_shifts": ["AI/ML adoption", "Cloud-first", "Mobile-first"],
                "market_drivers": ["Digital transformation", "Cost efficiency", "User experience"]
            }
        
        # Process funding data
        if "funding" in research_results:
            funding_data = research_results["funding"]
            compiled["funding_landscape"] = {
                "investor_interest": "High activity in sector",
                "funding_stages": ["Seed", "Series A", "Series B"],
                "average_funding": "$2-5M seed, $10-20M Series A"
            }
        
        # Generate summary
        compiled["summary"] = {
            "market_opportunity": f"Strong opportunity for {idea} in growing market",
            "key_challenges": ["Competition", "Market education", "Customer acquisition"],
            "success_factors": ["Product-market fit", "Strong team", "Scalable technology"],
            "recommendations": ["Focus on differentiation", "Build MVP quickly", "Validate with customers"]
        }
        
        return compiled
    
    def _extract_market_insights(self, market_data: Dict) -> List[str]:
        """Extract market insights from research data"""
        return [
            "Growing market demand",
            "Multiple customer segments",
            "Technology adoption increasing",
            "Investment activity high"
        ]
    
    def _extract_competitors(self, competitor_data: Dict) -> List[str]:
        """Extract competitor information from research data"""
        return [
            "Established enterprise players",
            "Emerging startup companies",
            "Traditional solution providers",
            "New technology entrants"
        ]
    
    def _extract_trends(self, trends_data: Dict) -> List[str]:
        """Extract trend information from research data"""
        return [
            "AI/ML integration",
            "Cloud-native solutions",
            "Mobile-first approach",
            "User experience focus"
        ]
    
    def _generate_fallback_search_result(self, query: str) -> Dict:
        """Generate fallback search result when API fails"""
        return {
            "query": query,
            "results": [
                {
                    "url": "https://example.com",
                    "title": f"Research data for {query}",
                    "content": f"Fallback research content for {query}. This would contain relevant market data and insights.",
                    "relevance": "medium"
                }
            ],
            "total_found": 1,
            "source": "fallback"
        }
    
    def _generate_fallback_research(self, research_type: str, idea: str) -> Dict:
        """Generate fallback research data for specific research type"""
        fallback_data = {
            "market_size": {
                "query": f"{idea} market size",
                "results": [{"content": f"The {idea} market shows strong growth potential with increasing demand."}],
                "source": "fallback"
            },
            "competitors": {
                "query": f"{idea} competitors",
                "results": [{"content": f"Competitive landscape for {idea} includes established players and emerging startups."}],
                "source": "fallback"
            },
            "trends": {
                "query": f"{idea} trends",
                "results": [{"content": f"Current trends in {idea} space include innovation and user-centric design."}],
                "source": "fallback"
            },
            "funding": {
                "query": f"{idea} funding",
                "results": [{"content": f"Investment activity in {idea} sector remains strong with multiple funding rounds."}],
                "source": "fallback"
            },
            "customers": {
                "query": f"{idea} customers",
                "results": [{"content": f"Target customers for {idea} include early adopters and enterprise users."}],
                "source": "fallback"
            }
        }
        
        return fallback_data.get(research_type, {
            "query": f"{idea} {research_type}",
            "results": [{"content": f"Fallback data for {research_type} research."}],
            "source": "fallback"
        })
    
    def _generate_fallback_comprehensive_research(self, idea: str, industry: Optional[str] = None) -> Dict:
        """Generate comprehensive fallback research when API is unavailable"""
        return {
            "idea": idea,
            "research_timestamp": datetime.now().isoformat(),
            "market_intelligence": {
                "size_indicators": ["Growing market", "Multiple segments", "High demand"],
                "growth_potential": "High based on industry trends",
                "key_segments": ["B2B", "B2C", "Enterprise"]
            },
            "competitive_landscape": {
                "direct_competitors": ["Established players", "Emerging startups", "Traditional solutions"],
                "competitive_advantages": ["Innovation", "Technology", "User Experience"],
                "market_gaps": ["Scalability", "Affordability", "Ease of use"]
            },
            "industry_trends": {
                "emerging_trends": ["AI/ML adoption", "Cloud-first", "Mobile-first"],
                "technological_shifts": ["Digital transformation", "Automation", "User-centric design"],
                "market_drivers": ["Cost efficiency", "Productivity", "Innovation"]
            },
            "funding_landscape": {
                "investor_interest": "High activity in technology sector",
                "funding_stages": ["Seed", "Series A", "Series B"],
                "average_funding": "$2-5M seed, $10-20M Series A"
            },
            "customer_insights": {
                "target_segments": ["Early adopters", "SME businesses", "Enterprise"],
                "pain_points": ["Cost", "Complexity", "Integration"],
                "adoption_factors": ["ROI", "Ease of use", "Support"]
            },
            "summary": {
                "market_opportunity": f"Strong opportunity for {idea} in growing market",
                "key_challenges": ["Competition", "Market education", "Customer acquisition"],
                "success_factors": ["Product-market fit", "Strong team", "Scalable technology"],
                "recommendations": ["Focus on differentiation", "Build MVP quickly", "Validate with customers"]
            },
            "source": "fallback_comprehensive"
        }